src/Dfc.CourseDirectory.Core/DataStore/Sql/QueryHandlers/GetTLevelsForProviderHandler.cs

```diff
@@ -0,0 +1,73 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Data.SqlClient;
+using System.Linq;
+using System.Threading.Tasks;
+using Dapper;
+using Dfc.CourseDirectory.Core.DataStore.Sql.Models;
+using Dfc.CourseDirectory.Core.DataStore.Sql.Queries;
+using Dfc.CourseDirectory.Core.Models;
+
+namespace Dfc.CourseDirectory.Core.DataStore.Sql.QueryHandlers
+{
+    public class GetTLevelsForProviderHandler :
+        ISqlQueryHandler<GetTLevelsForProvider, IReadOnlyCollection<TLevel>>
+    {
+        public async Task<IReadOnlyCollection<TLevel>> Execute(
+            SqlTransaction transaction,
+            GetTLevelsForProvider query)
+        {
+            var sql = @"
+SELECT
+    t.TLevelId, t.TLevelStatus, p.ProviderId, p.ProviderName,
+    t.WhoFor, t.EntryRequirements, t.WhatYoullLearn, t.HowYoullLearn, t.HowYoullBeAssessed,
+    t.WhatYouCanDoNext, t.YourReference, t.StartDate, t.Website,
+    d.TLevelDefinitionId, d.FrameworkCode, d.ProgType, d.Name
+FROM Pttcd.TLevels t
+JOIN Pttcd.TLevelDefinitions d ON t.TLevelDefinitionId = d.TLevelDefinitionId
+JOIN Pttcd.Providers p ON t.ProviderId = p.ProviderId
+WHERE t.ProviderId = @ProviderId AND t.TLevelStatus = @LiveTLevelStatus
+
+SELECT tll.TLevelId, tll.TLevelLocationId, tll.TLevelLocationStatus, tll.VenueId, v.VenueName
+FROM Pttcd.TLevelLocations tll
+JOIN Pttcd.TLevels t ON tll.TLevelId = t.TLevelId
+JOIN Pttcd.Venues v ON tll.VenueId = v.VenueId
+WHERE t.ProviderId = @ProviderId
+AND TLevelLocationStatus = @LiveTLevelLocationStatus";
+
+            var param = new
+            {
+                query.ProviderId,
+                LiveTLevelStatus = TLevelStatus.Live,
+                LiveTLevelLocationStatus = TLevelLocationStatus.Live
+            };
+
+            using (var reader = await transaction.Connection.QueryMultipleAsync(sql, param, transaction: transaction))
+            {
+                var tLevels = reader.Read<TLevel, TLevelDefinition, TLevel>(
+                    (tl, definition) =>
+                    {
+                        tl.TLevelDefinition = definition;
+                        return tl;
+                    },
+                    splitOn: "TLevelDefinitionId");
+
+                var tLevelLocations = reader.Read<TLevelLocationRow>()
+                    .ToLookup(r => r.TLevelId);
+
+                return tLevels
+                    .Select(tl =>
+                    {
+                        tl.Locations = tLevelLocations[tl.TLevelId].ToArray();
+                        return tl;
+                    })
+                    .ToArray();
+            }
+        }
+
+        private class TLevelLocationRow : TLevelLocation
```

Comment:
Data models should be representative of the data structure - this shouldn't be necessary if that's the case. If we want cleaner models to work with we should define them in the business layer and transform.

Reply:
I don't understand this. This is a convenience type for associating a location with its parent, specific to this query handler's implementation.

Reply:
Looking at this a little more closely I see what you're doing (creating custom data models populated from queries of multiple tables). Withdrawn.